### PR TITLE
fix: fn_param completion lookup

### LINKED ISSUE #12073: parameter completion are no longer being filtered by the client
```rs
fn foo(foo: u32, bar: u32, baz: u32, qux: u32, ) {}

fn bar(b$0)
```
shows all the parameters here, but the client should be able to filter all but `bar` and `baz` out.
image

### PR DESCRIPTION
**Pull request by rainy-me**
close #12073 caused by #12040

### CODE REVIEW

File: crates/ide_completion/src/completions/fn_param.rs [hunk 1]
```diff
@@ -56,7 +55,7 @@ fn fill_fn_params(
     ctx: &CompletionContext,
     function: &ast::Fn,
     param_list: &ast::ParamList,
-    mut add_new_item_to_acc: impl FnMut(&str, String),
+    mut add_new_item_to_acc: impl FnMut(&str),
 ) {
     let mut file_params = FxHashMap::default();
```
Comment (rainy-me) [inline crates/ide_completion/src/completions/fn_param.rs:60]: we still needs this to be a `FxHashMap` for removing duplicates

File: crates/ide_completion/src/completions/fn_param.rs [hunk 2]
```diff
@@ -161,16 +155,16 @@ fn should_add_self_completions(ctx: &CompletionContext, param_list: &ast::ParamL
     inside_impl && no_params
 }
 
-fn comma_wrapper(ctx: &CompletionContext) -> Option<(impl Fn(&str) -> String, TextRange, String)> {
+fn comma_wrapper(ctx: &CompletionContext) -> Option<(impl Fn(&str) -> String, TextRange)> {
     let param = ctx.token.ancestors().find(|node| node.kind() == SyntaxKind::PARAM)?;
-
+    let is_mut_token = (ctx.token.kind() == SyntaxKind::MUT_KW).then(|| &ctx.token);
```
Comment (rainy-me) [inline crates/ide_completion/src/completions/fn_param.rs:160]: I discovered this regression during testing: if current token is exactly `mut` kw, for example `fn foo(mut$0)`, the ancestor PARAM will become which include the R_PAREN and removes it incorrectly.
```
[crates/ide_completion/src/completions/fn_param.rs:160] &param = PARAM@27..31
  IDENT_PAT@27..31
    MUT_KW@27..30 "mut"
    ERROR@30..31
      R_PAREN@30..31 ")"
``` 
I decided just to special case it but not sure if there are more situations the ancestor PARAM range should be adjusted.

Comment (Veykril) [reply crates/ide_completion/src/completions/fn_param.rs:160]: Oh that is a bug in the parser I am pretty sure. Iirc we have the invariant that nodes have balanced parentheses which isn't the case here

Comment (Veykril) [reply crates/ide_completion/src/completions/fn_param.rs:160]: Should be fixed with https://github.com/rust-lang/rust-analyzer/pull/12090

**Comment by Veykril** [comment]
Thanks!
@bors r+

**Pull request merged**